### ISSUE TITLE
ros service compose format v2 support?

### ISSUE THREAD

**Issue by pwFoo** [rancher/os#1312]
**RancherOS Version: (ros os version)**
v0.7.0-rc4

**Where are you running RancherOS? (docker-machine, AWS, GCE, baremetal, etc.)**
notebook

Tried to add a service file with compose version v2

```
version: '2'
services:
  ...
```

Error message during sudo ros s build:
```
ERRO[0000] Could not parse config for project os : Unknown resolution for 'services' using <config.RawService Value> at line 1, column 18 
ERRO[0000] Failed to load /var/lib/rancher/conf/dengleros.yml: Unknown resolution for 'services' using <config.RawService Value> at line 1, column 18
```

So v2 isn't supported with RancherOS services? Need v2 because of compose tmpfs and group_add options.

Removed the first two lines and the service is created successful, but v2 tmpfs is missing.

**Comment by joshwget** [comment]
v2 format is currently only supported for external services. In that case the service files will look like standard Docker Compose files.

For cloud-config services I don't think we'll add the `version` key. Our current plan is to just detect the version for services based on the attributes used.

**Comment by joshwget** [comment]
I misread your original comment. It looks like you are loading the service externally. Can you post the full file and the command you used?

**Comment by pwFoo** [comment]
My current test files. enabled, build / create per local file 
```
sudo ros s enable /var/lib/rancher/conf/desktop.yml
...
```

*/var/lib/rancher/conf/desktop.yml*
```
#version: '2'
#services:
  xorg:
    image: "dengleros/xorg"
    container_name: xorg
    hostname: xorg
    privileged: true
    read_only: true
    restart: always
    environment:
    - DISPLAY=:0
    tmpfs:
    - /var/log
    volumes:
    - dengleros-tmp:/tmp
    # Workaround missing tmpfs
    - /var/log
    user: root

  dbus:
    image: "dengleros/dbus"
    container_name: dbus
    hostname: dbus
    read_only: true
    restart: always
    labels:                   
    - io.rancher.os.after=xorg
    volumes:
    - dengleros-varlibdbus:/var/lib/dbus
    - dengleros-varrundbus:/var/run/dbus
    - dengleros-homeuserdbus:/home/user/.dbus
    # xorg volume
    - dengleros-tmp:/tmp

  volumeicon:
    image: "dengleros/volumeicon"
    container_name: volumeicon
    hostname: volumeicon
    restart: always
    labels:
    - io.rancher.os.after=xorg
    environment:
    - DISPLAY=:0
    devices:
    - /dev/snd:/dev/snd
    volumes:
    - dengleros-volumeicon:/home/user/.config/volumeicon
    # Volume xorg
    - dengleros-tmp:/tmp 
```

**/var/lib/rancher/conf/fluxbox.yml**
```
#version: '2'
#services:
  fluxbox:                                   
    image: "dengleros/fluxbox"               
    container_name: fluxbox                  
    hostname: fluxbox                        
    restart: unless-stopped                  
    labels:                                  
    - io.rancher.os.after=xorg               
    - io.rancher.os.after=dbus               
    environment:                             
    - DISPLAY=:0                             
    volumes:                  
    - dengleros-homeuserfluxbox:/home/user/.fluxbox
    - /usr/bin/docker:/usr/bin/docker              
    - /var/run/docker.sock:/var/run/docker.sock    
    # Volumes xorg and dbus                        
    - dengleros-varlibdbus:/var/lib/dbus           
    - dengleros-varrundbus:/var/run/dbus           
    - dengleros-homeuserdbus:/home/user/.dbus      
    - dengleros-tmp:/tmp 
```

**/var/lib/rancher/conf/chromium.yml**
```
#version: '2'
#services:
  chromium:
    image: "dengleros/chromium"
    container_name: chromium
    hostname: chromium
    restart: on-failure
    labels:                   
    - io.rancher.os.after=xorg
    - io.rancher.os.after=dbus
    environment:
    - DISPLAY=:0
    group_add:
    - audio
    - video 
    volumes:               
    # Volumes xorg and dbus             
    - dengleros-varlibdbus:/var/lib/dbus
    - dengleros-varrundbus:/var/run/dbus     
    - dengleros-homeuserdbus:/home/user/.dbus
    - dengleros-tmp:/tmp
```

**Comment by joshwget** [comment]
Tested out a simple v2 file and I see the same error.

```
version: '2'
services:
  test:
    image: busybox
```

I guess our version of libcompose isn't new enough to support this yet. Once we update it then this should be supported for external services.

**Comment by pwFoo** [comment]
@joshwget 
Is there an easy way to update compose?
Need it to test my desktop build.

**Comment by pwFoo** [comment]
Tested 0.8.0-rc Release, but looks like v2 services still not supported.

**Comment by joshwget** [comment]
v2 support didn't make it into v0.8.0-rc1. This is unfortunately dependent on #1394, so there's actually a chance it might not make it into v0.8.0.

**Comment by pwFoo** [comment]
Ok, I'll pause my desktop build until v2 syntax is supported.

**Comment by vancluever** [comment]
This would be cool! Also, when this happens, can we make sure we get some support (or maybe earlier) for `ros` to be able to run services that are built off of the `build` directive in Compose files?

Right now it seems that we can build with `ros s build`, but starting the service after enabling the manually created YAML file quits silently. Not too sure but my assumption right now is due to the lack of an `image` directive since when that is added, I get an image not found unless it's `os_SERVICENAME`.

Use case: I wanted to use pure cloud-config to stage a Dockerfile with `write_files`, and then build and launch the service entirely with compose in `rancher -> services`. Eventually I got to this post where I discovered the compose format was v1, when lead me to this conclusion. It would be nice to have this feature, that way lightweight containers could be built and ran entirely through user data versus depending on a pre-built container, or having to roll them into your image.

**Comment by joshwget** [comment]
@vancluever That sounds like a pretty interesting use case. Would you mind filing a separate issue for it?

The current build command comes from libcompose. I suspect there's a bit of additional wiring on our end to make `ros` read from a Dockerfile and build context to make it work.

**Comment by vancluever** [comment]
Hey @joshwget for sure! I'll write one when I have a chance. Sorry for the delay in replying!

The fact that libcompose is used was basically what drove me to my initial assumptions that any old Compose directives would work as 100% intended - I'm not completely privy currently to how the actual compose product interfaces with Dockerfiles when `build:` is supplied but I hoped it would be pretty light.

Anyway will save the majority of the detail for the new issue :)

**Comment by pwFoo** [comment]
Because RancherOS have some new rc releases... Any progress with an updated compose support?
I would need shm_size, tmpfs and group_add for my desktop build as user-docker os-service :)

@SvenDowideit

**Comment by SvenDowideit** [comment]
@joshwget ?

**Comment by thebalaa** [comment]
+1

**Comment by mrflowo** [comment]
+1

**Comment by develar** [comment]
So disappointing. Any way, it seems there is easy workaround — http://rancher.com/docs/os/v1.1/en/configuration/running-commands/

**Comment by albertdb** [comment]
@niusmallnan Is this supported in v1.3.0?

**Comment by ravens** [comment]
Is there any plan to address this ? Compose v2 service definition would enable a richer set of usages, especially on the network definition side (external network, fixed IP address in macvlan...). Thanks !

**Comment by timorkal** [comment]
hello, any progress on that?
